src/backend/distributed/utils/metadata_cache.c

```diff
@@ -964,13 +975,245 @@ CitusHasBeenLoaded(void)
 			 * present during early stages of upgrade operation.
 			 */
 			DistPartitionRelationId();
+
+			/*
+			 * We also set citusCatalogVersion to NULL so that it will be re-read
+			 * in case of extension update.
+			 */
+			citusAvailableVersion = NULL;
+			citusCatalogVersion = NULL;
+		}
+	}
+
+	if (extensionLoaded)
+	{
+		if (!AvailableVersionMatchesBinaryVersion())
+		{
+			ereport(ERROR, (errmsg("citus binaries does not match available extension"),
+							(errdetail("Please restart the server"))));
+		}
+
+		if (!CatalogVersionMatchesBinaryVersion())
+		{
+			ereport(ERROR, (errmsg("citus binaries does not match installed extension"),
+							(errdetail("Please update installed extension with "
+									   "ALTER EXTENSION citus UPDATE;"))));
 		}
 	}
 
 	return extensionLoaded;
 }
 
 
+/*
+ * AvailableVersionMatchesBinaryVersion compares binary version of the Citus and
+ * currently available version from citus.control file. If they are same in major and
+ * minor version numbers , this function returns true. It ignores the patch version.
+ */
+static bool
+AvailableVersionMatchesBinaryVersion()
+{
+	char *availableVersion = CitusAvailableVersion();
+	const char patchSeparator = '-';
+
+	int comparisionLimit = strchr(availableVersion, patchSeparator) - availableVersion;
+
+	if (strncmp(availableVersion, CITUS_BINARY_VERSION, comparisionLimit) == 0)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * AvailableVersionMatchesBinaryVersion compares binary version of the Citus and
+ * catalog version from pg_extemsion catalog table. If they are same in major and
+ * minor version numbers , this function returns true. It ignores the patch version.
+ */
+static bool
+CatalogVersionMatchesBinaryVersion()
+{
+	char *catalogVersion = CitusCatalogVersion();
+	const char patchSeparator = '-';
+
+	int comparisionLimit = strchr(catalogVersion, patchSeparator) - catalogVersion;
+
+	if (strncmp(catalogVersion, CITUS_BINARY_VERSION, comparisionLimit) == 0)
+	{
+		return true;
+	}
+
+	return false;
```

Comment:
What do you think about flushing the cached variable if we're going to return `false`?

In particular, I'm wondering what happens if a user runs `ALTER EXTENSION` in one connection. Other connections won't know yet and will keep erroring.